File: LC/Python/Easy/110.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root: return True
        balanced = [True]
        
        def dfs(curr):
            if not balanced[0]:
                return False
            
            if not curr:
                return 0
            
            left = dfs(curr.left)
            right = dfs(curr.right)
            
            if abs(left - right) > 1:
                balanced[0] = False
            
            return max(left, right) + 1
        
        dfs(root)
        return balanced[0]

File: LC/C++/Easy/110.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool ans = true;

    int dfs(TreeNode* curr) {
        if (curr == nullptr || ans == false) { return 0; }
        int left = dfs(curr->left);
        int right = dfs(curr->right);
        if (abs(left - right) > 1) {
            ans = false;
            return 0;
        }

        return max(left, right) + 1;
    }

    bool isBalanced(TreeNode* root) {
        dfs(root);
        return ans;
    }
};